esp/GBPlay/main/hardware/wifi.c

```diff
@@ -8,109 +9,60 @@
 #include "wifi.h"
 
 #define MAX_CONNECTION_RETRY_COUNT 3
+#define CONNECTION_TIMEOUT_MS      15000
 
 #define WIFI_SAVED_NETWORKS_STORAGE_KEY "wifi_networks"
 
 typedef enum {
     CONNECTION_FAIL    = 1,
     CONNECTION_SUCCESS = 2
-} WifiConnectionStatus;
+} WifiConnectionEvent;
 
 typedef struct {
     wifi_network_credentials networks[WIFI_MAX_SAVED_NETWORKS];
     int count;
 } wifi_saved_network_info;
 
+static SemaphoreHandle_t s_wifi_lock;
+static SemaphoreHandle_t s_wifi_storage_lock;
+static EventGroupHandle_t s_wifi_event_group;  // For blocking on connect
 static wifi_saved_network_info s_saved_networks = {0};
-
-static volatile bool is_connected = false;
-
-static esp_netif_t* default_iface = NULL;
-
-// FreeRTOS event group to signal when we are connected
-static EventGroupHandle_t wifi_event_group;
-static int connection_retry_count = 0;
+static esp_netif_t* s_wifi_iface = NULL;
+static volatile bool s_is_connected = false;
 
 static void _set_connection_status(bool connected)
 {
-    is_connected = connected;
+    s_is_connected = connected;
 }
 
-static void _on_wifi_event(void* arg, esp_event_base_t event_base, int32_t event_id, void* event_data)
+static void _on_disconnect(void* arg, esp_event_base_t event_base, int32_t event_id, void* event_data)
 {
-    switch (event_id)
-    {
-        case WIFI_EVENT_STA_START:
-            ESP_LOGI(__func__, "Wi-Fi started");
-            break;
-        case WIFI_EVENT_STA_CONNECTED:
-            ESP_LOGI(__func__, "Wi-Fi connected");
-            break;
-        case WIFI_EVENT_STA_DISCONNECTED:
-        {
-            wifi_event_sta_disconnected_t* event = (wifi_event_sta_disconnected_t*)event_data;
-
-            _set_connection_status(false);
+    wifi_event_sta_disconnected_t* event = (wifi_event_sta_disconnected_t*)event_data;
+    ESP_LOGI(__func__, "Disconnected from network %s", event->ssid);
 
-            if (event->reason == WIFI_REASON_ASSOC_LEAVE)
-            {
-                // Left voluntarily
-                ESP_LOGI(__func__, "Left Wi-Fi network %s", event->ssid);
-            }
-            else if (connection_retry_count <= MAX_CONNECTION_RETRY_COUNT)
-            {
-                // Failed to connect, or AP went away
-                ESP_LOGI(__func__, "Wi-Fi disconnected. Retrying connection to %s", event->ssid);
-                ++connection_retry_count;
-                esp_wifi_connect();
```

Comment:
Retrying on connect is now handled in `wifi_connect()` so we don't do it if we disconnect at other times (that will be the wifi manager's responsibility)